otc.go

```diff
@@ -0,0 +1,240 @@
+	CredentialKeyUsername   = "username"
+	CredentialKeyPassword   = "password"
+	identityEndpointFormat  = "https://iam.%s.otc.t-systems.com/v3"
+)
+
+var (
+	ErrZoneNotFound   = errors.New("could not find zone by the given name")
+	ErrRecordNotFound = errors.New("could not find record by the given name")
+)
+
+func NewOtcProvider(_ context.Context, _ string, credentialsData map[string]string, logger api.Logger) (*OTC, error) {
+	for _, key := range []string{CredentialKeyRegion, CredentialKeyDomainName, CredentialKeyTenantName, CredentialKeyUsername, CredentialKeyPassword} {
+		if _, isSet := credentialsData[key]; !isSet {
+			return nil, fmt.Errorf("missing key %s is credentialData", key)
+		}
+	}
+
+	client, err := openstack.AuthenticatedClient(golangsdk.AuthOptions{
+		IdentityEndpoint: fmt.Sprintf(identityEndpointFormat, credentialsData[CredentialKeyRegion]),
+		DomainName:       credentialsData[CredentialKeyDomainName],
+		TenantName:       credentialsData[CredentialKeyTenantName],
+		Username:         credentialsData[CredentialKeyUsername],
+		Password:         credentialsData[CredentialKeyPassword],
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize authenticated client: %v", err)
+	}
+
+	dns, err := openstack.NewDNSV2(client, golangsdk.EndpointOpts{
+		Region: credentialsData[CredentialKeyRegion],
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to init dns client: %v", err)
+	}
+
+	return &OTC{
+		client: client,
+		dns:    dns,
+		region: credentialsData[CredentialKeyRegion],
+		logger: logger,
+	}, nil
+}
+
+func (o OTC) GetDNSRecords(ctx context.Context, zone, name string) ([]api.Record, error) {
+	z, err := o.findZoneByName(ctx, zone)
+	if err != nil {
+		return nil, err
+	}
+
+	zoneID := z.ID
+
+	recordSets, err := o.listRecordSets(ctx, zoneID, name, "")
+	if err != nil {
+		return nil, err
+	}
+
+	var apiRecords []api.Record
+	for _, rec := range recordSets {
+		fqdn := fmt.Sprintf("%s.%s", name, zone)
+		if name == "" || rec.Name == fqdn {
+			apiRecords = append(apiRecords, api.Record{
+				Type:    rec.Type,
+				Name:    name,
+				Content: rec.Records,
+				TTL:     rec.TTL,
+			})
+		}
+	}
+
+	return apiRecords, nil
+}
+
+func (o OTC) CreateOrUpdateDNSRecord(ctx context.Context, zone, name, rtype, content string, ttl int, proxy bool) error {
+	z, err := o.findZoneByName(ctx, zone)
+	if err != nil {
+		return err
+	}
+
+	zoneID := z.ID
+
+	records, err := o.listRecordSets(ctx, zoneID, name, rtype)
+	if err != nil {
+		return err
+	}
+
+	if len(records) == 0 {
+		if err := o.createDNSRecord(ctx, zoneID, fmt.Sprintf("%s.%s", name, zone), rtype, content, ttl, proxy); err != nil {
+			return fmt.Errorf("failed to create record in  zoneID '%s' (zone name '%s') with name %s: %v", zoneID, zone, name, err)
+		}
+
+		return nil
+	}
+
+	record := records[0]
+
+	// no change
+	if record.TTL == ttl && record.Records[0] == strings.Trim(content, "\"") {
+		return nil
+	}
```

Comment:
If this is a TXT record, would record.Records[0] be quoted? If so that equality check will fail.  You may need to strip quotes on both before comparing.

Reply:
That already happens in `listRecordSets` in line `217`.

Reply:
Ah sorry, I missed that.